File: api/src/main/java/net/socket/FrogClientSocket.java
package net.socket;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.ArrayList;
import java.util.List;

import org.json.JSONObject;

import net.IPacketListener;
import net.Packet;
import net.PacketReaderWorker;
import net.PacketSubscriber;
import net.PacketType;
import utils.FrogException;

/**
 * <h1>Gestionnaire de packets internet de jeu</h1>
 * <p>Classe permettant de gérer la communication avec le serveur de jeu</p>
 */
public class FrogClientSocket implements IPacketListener
{
	public static final String PROTOCOL_VERSION = "v0.0.0.1";
	
	private static final int DURATION_BETWEEN_TENTATIVES = 1000;
	private static final int TENTATIVES = 10;
	
	private List<PacketSubscriber> balancers;
	private Socket socket;
	private String token;
	private boolean isRunning;
	private BufferedReader reader;
	private PrintWriter writer;
	
	public FrogClientSocket()
	{
		this.balancers = new ArrayList<PacketSubscriber>();
		this.isRunning = false;
	}
	
	/**
	 * <h1>Ajoute un nouveau aiguilleur de packets</h1>
	 * <p>Les souscripteurs de l'aiguilleur recevront les packets en reception</p>
	 * @param balancer Le nouveau aiguilleur à ajouter
	 */
	public synchronized void addPacketSubscribers(PacketSubscriber balancer)
	{
		this.balancers.add(balancer);
	}
	
	/**
	 * <h1>Supprime un aiguilleur de packets</h1>
	 * <p>Les souscripteurs de l'aiguilleur ne recevront plus les packets en reception</p>
	 * @param balancer L'aiguilleur à supprimer
	 */
	public synchronized void removePacketSubscribers(PacketSubscriber balancer)
	{
		this.balancers.remove(balancer);
	}
	
	/**
	 * Démarre la communication avec le serveur de jeu si possible
	 * @param ip L'adresse du serveur hôte
	 * @param port Le port d'écoute du serveur hôte
	 * @throws FrogException
	 */
	public void start(String ip, int port) throws FrogException
	{
		if(this.isRunning())
			return;
		
		if(this.trySocketConnect(ip, port, TENTATIVES, DURATION_BETWEEN_TENTATIVES))
		{
			this.isRunning = true;
			PacketReaderWorker worker = new PacketReaderWorker(this);
			worker.addListener(this);
			worker.start(); // Lancement du thread de lecture des packets en reception
		}
	}
	
	@Override
	public void onPacketReceived(Packet packet) 
	{
		switch(packet.getType())
		{
			case CONNECT_RESULT:
				if(this.token == null)
				{
					JSONObject obj = new JSONObject(packet.getSerializedObject());
					boolean result = obj.getBoolean("result");
					if(result)
						this.token = obj.getString("token");
				}
				break;
			default:
				break;
		}
		
		
		this.raiseEventToBalancers(packet);
		
		// do something with received packet...
	}

	/**
	 * Obtient l'état de la communication avec le serveur de jeu
	 * @category Couche serveur de jeu
	 * @return True si la communication est active, False dans le cas contraire
	 */
	public synchronized boolean isRunning()
	{
		return this.isRunning;
	}
	
	/**
	 * Défini l'arrêt de la communication avec le serveur de jeu
	 */
	public synchronized void stop()
	{
		this.isRunning = false;
	}
	
	/**
	 * Obtient le gestionnaire de flux en entrée
	 * @return Le gestionnaire de flux en entrée
	 */
	public BufferedReader getReader()
	{
		return this.reader;
	}
	
	/**
	 * Obtient le gestionnaire de flux en sortie
	 * @return Le gestionnaire de flux en sortie
	 */
	public PrintWriter getWriter()
	{
		return this.writer;
	}
	
	/**
	 * <h1>Tentative de connection avec l'hôte distant</h1>
	 * <p>Tente une connexion avec l'hôte distant spécifié
	 * Si la connexion échoue, une nouvelle tentative est relancé au bout d'un temps défini et celà
	 * un nombre de fois défini également.</p>
	 * @param ip L'IP du serveur hôte
	 * @param port Le PORT d'écoute du serveur hôte
	 * @param tentativesCount Le nombre de tentatives de connexion maximum
	 * @param milisecondsBetweenEachTentatives Le temps d'attente entre chaque tentative en milisecondes
	 * @return True si la connexion a été réalisée avec succès, False dans le cas contraire
	 * @throws FrogException Jetée lorsqu'une erreur système survient
	 */
	private boolean trySocketConnect(String ip, int port, int tentativesCount, int milisecondsBetweenEachTentatives) throws FrogException
	{
		boolean result = false;
		int tentatives = 0;
		
		try
		{
			while(!result && tentatives <= tentativesCount)
			{
				try
				{
					this.socket = new Socket(ip, port);
					result = true;
				} 
				catch (Exception e) 
				{
					System.err.println(e.getMessage());
					result = false;
					tentatives++;
					Thread.sleep(milisecondsBetweenEachTentatives);
				}
			}
			
			if(result)
			{
				this.reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
				this.writer = new PrintWriter(this.socket.getOutputStream(), true);
				
				JSONObject obj = new JSONObject();
				obj.put("version", PROTOCOL_VERSION);
				this.sendPacket(new Packet(PacketType.PROTOCOL_VERSION, obj.toString()));
				Packet p = Packet.getPacket(this.reader.readLine());
				
				if(p.getType() != PacketType.PROTOCOL_VERSION_RESULT)
					throw new FrogException("Mauvais packet reçu; Attendu : PROTOCOL_VERSION_RESULT");
				
				JSONObject versionPacket = new JSONObject(p.getSerializedObject());
				boolean versionResult = versionPacket.getBoolean("result");
				
				if(!versionResult)
				{
					this.isRunning = false;
					throw new FrogException("Mauvaise version du protocol de communication");
				}
			}
		}
		catch(Exception e)
		{
			throw new FrogException(e.getMessage());
		}
		
		return result;
	}
	
	/**
	 * <h1>Tentative de connexion au serveur de jeu</h1>
	 * <p>Tente une connexion au serveur de jeu avec les informations d'authentification spécifiées.</p>
	 * @param account Le nom de compte joueur
	 * @param password Le mot de passe du compte joueur
	 * @throws FrogException
	 */
	public void connect(String account, String password) throws FrogException
	{
		this.sendPacket(new Packet(PacketType.CONNECT, "Nothing"));
	}
	
	public void sendPacket(Packet packet) throws FrogException
	{
		if(!this.socket.isConnected())
			throw new FrogException("Tentative de connection au serveur de jeu sur un channel de communication non établi");
		
		writer.println(packet.toJSON());
	}
	
	/**
	 * Obtient l'état de la communication avec l'hôte
	 * @category Couche socket
	 * @return True si la communication est active, False dans le cas contraire
	 */
	public synchronized boolean isConnected()
	{
		if(this.socket != null && !this.socket.isClosed())
			return this.socket.isConnected();
		return false;
	}
	
	/**
	 * Obtient le token de communication avec le serveur de jeu résultant de l'authentification
	 * @return Null si l'authentification n'a pas été réalisée, ou a échouée, une clé dans le cas contraire
	 */
	public String getToken() 
	{
		return token;
	}
	
	private void raiseEventToBalancers(Packet packet)
	{
		for(int i = 0; i < this.balancers.size(); i++)
		{
			this.balancers.get(i).pushPacket(packet);
		}
	}
}

File: godot/scripts/api/net/socket/FrogClientSocket.cs
using System.Net.Sockets;
using System.Text.Json;
using FrogCreator.Api.Utils;

namespace FrogCreator.Api.Net.Socket;

/// <summary>
/// Gestionnaire de packets internet de jeu.
/// Classe permettant de gérer la communication avec le serveur de jeu.
/// </summary>
public class FrogClientSocket : IPacketListener
{
    public static readonly string PROTOCOL_VERSION = "v0.0.0.1";

    private static readonly int DURATION_BETWEEN_TENTATIVES = 1000;
    private static readonly int TENTATIVES = 10;

    private readonly List<PacketSubscriber> _balancers;
    private TcpClient? _socket;
    private string? _token;
    private bool _isRunning;
    private StreamReader? _reader;
    private StreamWriter? _writer;
    private readonly object _lock = new object();

    public FrogClientSocket()
    {
        _balancers = new List<PacketSubscriber>();
        _isRunning = false;
    }

    /// <summary>
    /// Ajoute un nouveau aiguilleur de packets.
    /// Les souscripteurs de l'aiguilleur recevront les packets en reception.
    /// </summary>
    /// <param name="balancer">Le nouveau aiguilleur à ajouter</param>
    public void AddPacketSubscribers(PacketSubscriber balancer)
    {
        lock (_lock)
        {
            _balancers.Add(balancer);
        }
    }

    /// <summary>
    /// Supprime un aiguilleur de packets.
    /// Les souscripteurs de l'aiguilleur ne recevront plus les packets en reception.
    /// </summary>
    /// <param name="balancer">L'aiguilleur à supprimer</param>
    public void RemovePacketSubscribers(PacketSubscriber balancer)
    {
        lock (_lock)
        {
            _balancers.Remove(balancer);
        }
    }

    /// <summary>
    /// Démarre la communication avec le serveur de jeu si possible
    /// </summary>
    /// <param name="ip">L'adresse du serveur hôte</param>
    /// <param name="port">Le port d'écoute du serveur hôte</param>
    public void Start(string ip, int port)
    {
        if (IsRunning())
            return;

        if (TrySocketConnect(ip, port, TENTATIVES, DURATION_BETWEEN_TENTATIVES))
        {
            _isRunning = true;
            PacketReaderWorker worker = new PacketReaderWorker(this);
            worker.AddListener(this);
            worker.Start(); // Lancement du thread de lecture des packets en reception
        }
    }

    public void OnPacketReceived(Packet packet)
    {
        switch (packet.GetPacketType())
        {
            case PacketType.CONNECT_RESULT:
                if (_token == null)
                {
                    using JsonDocument doc = JsonDocument.Parse(packet.GetSerializedObject());
                    JsonElement root = doc.RootElement;
                    bool result = root.GetProperty("result").GetBoolean();
                    if (result)
                        _token = root.GetProperty("token").GetString();
                }
                break;
            default:
                break;
        }

        RaiseEventToBalancers(packet);
    }

    /// <summary>
    /// Obtient l'état de la communication avec le serveur de jeu
    /// </summary>
    /// <returns>True si la communication est active, False dans le cas contraire</returns>
    public bool IsRunning()
    {
        lock (_lock)
        {
            return _isRunning;
        }
    }

    /// <summary>
    /// Défini l'arrêt de la communication avec le serveur de jeu
    /// </summary>
    public void Stop()
    {
        lock (_lock)
        {
            _isRunning = false;
        }
    }

    /// <summary>
    /// Obtient le gestionnaire de flux en entrée
    /// </summary>
    public StreamReader? GetReader()
    {
        return _reader;
    }

    /// <summary>
    /// Obtient le gestionnaire de flux en sortie
    /// </summary>
    public StreamWriter? GetWriter()
    {
        return _writer;
    }

    /// <summary>
    /// Tentative de connection avec l'hôte distant.
    /// Tente une connexion avec l'hôte distant spécifié.
    /// Si la connexion échoue, une nouvelle tentative est relancé au bout d'un temps défini
    /// et celà un nombre de fois défini également.
    /// </summary>
    /// <param name="ip">L'IP du serveur hôte</param>
    /// <param name="port">Le PORT d'écoute du serveur hôte</param>
    /// <param name="tentativesCount">Le nombre de tentatives de connexion maximum</param>
    /// <param name="millisecondsBetweenEachTentatives">Le temps d'attente entre chaque tentative en milisecondes</param>
    /// <returns>True si la connexion a été réalisée avec succès, False dans le cas contraire</returns>
    /// <exception cref="FrogException">Jetée lorsqu'une erreur système survient</exception>
    private bool TrySocketConnect(string ip, int port, int tentativesCount, int millisecondsBetweenEachTentatives)
    {
        bool result = false;
        int tentatives = 0;

        try
        {
            while (!result && tentatives <= tentativesCount)
            {
                try
                {
                    _socket = new TcpClient(ip, port);
                    result = true;
                }
                catch (Exception e)
                {
                    System.Console.Error.WriteLine(e.Message);
                    result = false;
                    tentatives++;
                    Thread.Sleep(millisecondsBetweenEachTentatives);
                }
            }

            if (result && _socket != null)
            {
                NetworkStream stream = _socket.GetStream();
                _reader = new StreamReader(stream);
                _writer = new StreamWriter(stream) { AutoFlush = true };

                var versionObj = new Dictionary<string, object> { { "version", PROTOCOL_VERSION } };
                SendPacket(new Packet(PacketType.PROTOCOL_VERSION, JsonSerializer.Serialize(versionObj)));
                string? responseLine = _reader.ReadLine();
                if (responseLine == null)
                    throw new FrogException("Pas de réponse du serveur");

                Packet p = Packet.GetPacket(responseLine);

                if (p.GetPacketType() != PacketType.PROTOCOL_VERSION_RESULT)
                    throw new FrogException("Mauvais packet reçu; Attendu : PROTOCOL_VERSION_RESULT");

                using JsonDocument versionDoc = JsonDocument.Parse(p.GetSerializedObject());
                bool versionResult = versionDoc.RootElement.GetProperty("result").GetBoolean();

                if (!versionResult)
                {
                    _isRunning = false;
                    throw new FrogException("Mauvaise version du protocol de communication");
                }
            }
        }
        catch (FrogException)
        {
            throw;
        }
        catch (Exception e)
        {
            throw new FrogException(e.Message);
        }

        return result;
    }

    /// <summary>
    /// Tentative de connexion au serveur de jeu.
    /// Tente une connexion au serveur de jeu avec les informations d'authentification spécifiées.
    /// </summary>
    /// <param name="account">Le nom de compte joueur</param>
    /// <param name="password">Le mot de passe du compte joueur</param>
    public void Connect(string account, string password)
    {
        SendPacket(new Packet(PacketType.CONNECT, "Nothing"));
    }

    public void SendPacket(Packet packet)
    {
        if (_socket == null || !_socket.Connected)
            throw new FrogException("Tentative de connection au serveur de jeu sur un channel de communication non établi");

        _writer?.WriteLine(packet.ToJSON());
    }

    /// <summary>
    /// Obtient l'état de la communication avec l'hôte
    /// </summary>
    /// <returns>True si la communication est active, False dans le cas contraire</returns>
    public bool IsConnected()
    {
        lock (_lock)
        {
            if (_socket != null)
                return _socket.Connected;
            return false;
        }
    }

    /// <summary>
    /// Obtient le token de communication avec le serveur de jeu résultant de l'authentification
    /// </summary>
    /// <returns>Null si l'authentification n'a pas été réalisée, ou a échouée, une clé dans le cas contraire</returns>
    public string? GetToken()
    {
        return _token;
    }

    private void RaiseEventToBalancers(Packet packet)
    {
        for (int i = 0; i < _balancers.Count; i++)
        {
            _balancers[i].PushPacket(packet);
        }
    }
}
